Insert content text input client breaking change (#7692)
flutter/website: src/release/breaking-changes/insert-content-text-input-client.md

---
title: Insert Content Text Input Client
description: Add a new method to the TextInputClient interface to allow Android virtual keyboards to
insert rich content into Flutter TextFields. 
---

## Summary

Adds `insertContent` method to the `TextInputClient` interface to allow Android's image keyboard 
feature to insert content into a Flutter `TextField`.

## Context

> As of Android 7.1, IMEs (virtual keyboards) are able to directly send images and rich content
> into a text editor. This allows users to directly insert gifs, stickers, or context-aware rich
> content into a text field.

Reviewer: **issue:** Tighten language. (Removed adverbs as extraneous.)
```suggestion
As of Android 7.1, IMEs (input method editors or virtual keyboards) can send
images and rich content into a text editor. This allows users to insert gifs,
stickers, or context-aware rich content into a text field.
```

## Description of change

When the user inserts rich content via the IME, the platform sends a `commitContent` channel message
called `TextInputClient.commitContent`.

> This notifies the Dart code that rich content was inserted. 

Reviewer: **issue:** Fix passive voice.
```suggestion
This notifies the Dart code that the IME inserted rich content. 
```

The channel message contains the mime type, URI, and bytedata for the inserted content in JSON form.

## Migration guide

> If you previously implemented `TextInputClient`, you must override

Reviewer: **issue:** Remove adverb.
```suggestion
If you had implemented `TextInputClient`, you must override
```

`insertContent` to either support rich content insertion or provide an empty implementation.

To migrate, implement `insertContent`.

Code before migration:

<!-- skip -->
```dart
class MyCustomTextInputClient implements TextInputClient {
  ...
}
```

Code after migration:

<!-- skip -->
```dart
class MyCustomTextInputClient implements TextInputClient {
  ...
  @override
  void insertContent() {
    ...
  }
  ...
}

> ```

Reviewer 2: Some ideas of other things we might want to add:

  1. Maybe this is obvious, but it might help to state that if their implementation of TextInputClient doesn't need the ability to receive image data from Android virtual keyboards, then they can feel free to leave the implementation empty.

```dart
  @override
  void insertContent() {}
```
  2. State that instead of adding an empty implementation, if they don't want to be broken by changes like this in the future, they can use TextInputClient with the `with` keyword instead of `implements`.
  3. Maybe we should direct them to the implementation in TextInputClient as an example.  Like "TextInputClient calls simply calls a callback when `insertContent` is called.  See <link to code on GitHub when it's merged> for more details."

---
title: Insert Content Text Input Client
description: Add a new method to the TextInputClient interface to allow Android virtual keyboards to insert rich content into Flutter TextFields. 
---

## Summary

Added an `insertContent` method to the `TextInputClient` interface to allow Android's image keyboard
feature to insert content into a Flutter `TextField`.

## Context

As of Android 7.1, IMEs (input method editors or virtual keyboards) can send
images and rich content into a text editor. This allows users to insert gifs,
stickers, or context-aware rich content into a text field.

## Description of change

When the user inserts rich content in the IME, the platform sends a `TextInputClient.commitContent`
channel message, notifying the Dart code that the IME inserted rich content.
The channel message contains the mime type, URI, and bytedata for the inserted content in JSON form.

## Migration guide

> If you previously implemented the `TextInputClient` interface, you must override
> `insertContent` to either support rich content insertion or provide an empty implementation.

Reviewer: **nit:** Rephrase to be more concise.
```suggestion
If you previously the `TextInputClient` interface earlier, override
`insertContent` to either support rich content insertion or provide an empty implementation.
```

To migrate, implement `insertContent`.

Code before migration:

<!-- skip -->
```dart
class MyCustomTextInputClient implements TextInputClient {
  // ...
}
```

Code after migration:

<!-- skip -->
```dart
class MyCustomTextInputClient implements TextInputClient {
  // ...
  @override
  void insertContent() {
    // ...
  }
  // ...
}
```

> If your implementation of `TextInputClient` doesn't require the ability to receive rich content
> inserted from the IME, you can leave the implementation of `insertContent` empty with no
> consequences.

Reviewer: **suggestion:** Rephrase to make more concise.
```suggestion
Your implementation of `TextInputClient` might not require the ability to receive rich content
inserted from the IME. If that's the case, you can leave the implementation of `insertContent`
empty with no consequences.
```

<!-- skip -->
```dart
class MyCustomTextInputClient implements TextInputClient {
  // ...
  @override
  void insertContent() {}
  // ...
}
```

As an alternative, you can use a similar implementation to the default `TextInputClient`.

> To learn how to do this, see the [insertContent implementation]({{site.master-api}}/flutter/services/TextInputClient/insertContent.html).

Reviewer 3: ```suggestion
To learn how to do this, see the [insertContent implementation]({{site.api}}/flutter/services/TextInputClient/insertContent.html).
```

> Alternatively, you can use a similar implementation to the default `TextInputClient`.<br>
> For more details, see the [insertContent implementation]({{site.master-api}}/flutter/services/TextInputClient/insertContent.html) for more details.

Reviewer: **issue:** Redundant copy.
```suggestion
As an alternative, you can use a similar implementation to the default `TextInputClient`.
To learn how to do this, see the [insertContent implementation]({{site.master-api}}/flutter/services/TextInputClient/insertContent.html).
```

> In the future, you can prevent being broken by changes to an interface
> by using `with TextInputClient` rather than `implements TextInputClient`.

Reviewer: **issue:** Too much passive voice. Please revise.
```suggestion
To prevent breaking changes to an interface,
use `with TextInputClient` rather than `implements TextInputClient`.
```

## Timeline

Landed in version: 3.8.0-1.0.pre<br>

> In stable release: TBD

Reviewer 3: Looks like this landed in 3.10.0:

```suggestion
In stable release: 3.10.0
```

---
title: Insert Content Text Input Client
description: Add a new method to the TextInputClient interface to allow Android virtual keyboards to insert rich content into Flutter TextFields. 
---

## Summary

Added an `insertContent` method to the `TextInputClient` interface to allow Android's image keyboard 
feature to insert content into a Flutter `TextField`.

> 

Reviewer 4: Could you make a format operation to the document? I saw there are multiple blank lines that don't follow the regular behavior in this repo.

Author: Is this some sort of command I need to run in the terminal? Sorry, not super familiar with what tools are available when creating markdown docs.

Reviewer 4: Maybe a format action from editors? Such as VS Code and IDEA, or other markdown editors should support formatting too.

## Context

As of Android 7.1, IMEs (input method editors or virtual keyboards) can send
images and rich content into a text editor. This allows users to insert gifs,
stickers, or context-aware rich content into a text field.

## Description of change

When the user inserts rich content in the IME, the platform sends a `TextInputClient.commitContent`
channel message, notifying the Dart code that the IME inserted rich content.
The channel message contains the mime type, URI, and bytedata for the inserted content in JSON form.

## Migration guide

If you previously implemented the `TextInputClient` interface, you must override
`insertContent` to either support rich content insertion or provide an empty implementation.

To migrate, implement `insertContent`.

Code before migration:

<!-- skip -->
```dart
class MyCustomTextInputClient implements TextInputClient {
  ...
}
```

Code after migration:

<!-- skip -->
```dart
class MyCustomTextInputClient implements TextInputClient {

>   ...
>   @override
>   void insertContent() {
>     ...
>   }
>   ...

Reviewer 4: This makes it easy to copy-paste directly.

```suggestion
  // ...
  @override
  void insertContent() {
    // Implement insertContent.
  }
  // ...
```

---

> title: Insert Content Text Input Client
> description: Add a new method to the TextInputClient interface to allow Android virtual keyboards to
> insert rich content into Flutter TextFields. 

Reviewer 5: The description field in the yaml matter must be on one line.

```suggestion
title: Insert Content Text Input Client
description: Add a new method to the TextInputClient interface to allow Android virtual keyboards to insert rich content into Flutter TextFields. 
```

---

## Summary

> Adds `insertContent` method to the `TextInputClient` interface to allow Android's image keyboard 
> feature to insert content into a Flutter `TextField`.

Reviewer 5: Needs to be written in past tense.

```suggestion
Added an `insertContent` method to the `TextInputClient` interface to allow Android's image keyboard 
feature to insert content into a Flutter `TextField`.
```

## Context

As of Android 7.1, IMEs (input method editors or virtual keyboards) can send
images and rich content into a text editor. This allows users to insert gifs,
stickers, or context-aware rich content into a text field.

## Description of change

> When the user inserts rich content via the IME, the platform sends a `commitContent` channel message
> called `TextInputClient.commitContent`.
> This notifies the Dart code that the IME inserted rich content.
> The channel message contains the mime type, URI, and bytedata for the inserted content in JSON form.

Reviewer 5: ```suggestion
When the user inserts rich content in the IME, the platform sends
a `TextInputClient.commitContent` channel message, notifying
the Dart code that the IME inserted rich content.
The channel message contains the mime type, URI, and bytedata for the inserted content in JSON form.
```

## Migration guide

> If you had implemented `TextInputClient`, you must override
> `insertContent` to either support rich content insertion or provide an empty implementation.

Reviewer 5: If they provide an empty implementation, what happens? The rich content is
still added, yes?

```suggestion
If you previously implemented the `TextInputClient` interface, you must override
`insertContent` to either support rich content insertion or provide an empty implementation.
```

To migrate, implement `insertContent`.

Code before migration:

<!-- skip -->
```dart
class MyCustomTextInputClient implements TextInputClient {
  ...
}
```

Code after migration:

<!-- skip -->
```dart
class MyCustomTextInputClient implements TextInputClient {
  ...
  @override
  void insertContent() {
    ...
  }
  ...
}
```

> If your implementation of `TextInputClient` does not require the ability to receive rich content 
> inserted from the IME, you may leave the implementation of `insertContent` empty with no 
> consequences.

Reviewer 5: I guess this answers my question above. An empty implementation has no consequences. But what does that mean, exactly? The rich content is still added?

```suggestion
If your implementation of `TextInputClient` doesn't require the ability
to receive rich content inserted from the IME, you can leave the implementation
of `insertContent` empty with no  consequences.
```

<!-- skip -->
```dart
class MyCustomTextInputClient implements TextInputClient {
  ...
  @override
  void insertContent() {}
  ...
}
```

> Alternatively, you may use a similar implementation to the default `TextInputClient`.<br>
> See the [insertContent implementation]({{site.repo.flutter}}/TBD) for more details.
> 
> In the future, you can prevent being broken by changes like this by using `with TextInputClient`
> rather than `implements TextInputClient`.

Reviewer 5: I can't land this with a TBD link. Do you have an actual link?

```suggestion
Alternatively, you can use a similar implementation to the default `TextInputClient`.
For more details, see the [insertContent implementation]({{site.repo.flutter}}/TBD).

In the future, you can prevent being broken by changes to an interface
by using `with TextInputClient` rather than `implements TextInputClient`.
```

## Timeline

> Landed in version: TBD<br>

Reviewer 5: Again, I can't land this with info on where it landed. The stable release field can be TBD for now.
